htsohm/binning.py

```diff
@@ -36,44 +13,32 @@ def select_parents(run_id, children_per_generation, generation):
     for new materials, because their children are most likely to display unique properties. This
     function first calculates a `weight` for each bin, based on the number of constituent
     materials. These weights affect the probability of selecting a parent from that bin. Once a bin
-    is selected, a parent is randomly-selected from those materials within that bin."""
-
-    # Each bin is counted, then assigned a weight
-    config = read_config_file(run_id)
-    bins = config["number-of-bins"]
-    counts = count_all(run_id)
-    weights = np.zeros([bins, bins, bins])
-    for i in range(bins):
-        for j in range(bins):
-            for k in range(bins):
-                if counts[i,j,k] != 0.:
-                    weights[i,j,k] = counts.sum() / counts[i,j,k]
-    weights = weights / weights.sum()
-
-    ############################################################################
-    # The 3-dimensional list of weights is converted into a 1-dimensional list,
-    # and another 1-dimensional list of material-IDs is also generated.
-    weight_list = []
-    id_list = []
-    for i in range(bins):
-        for j in range(bins):
-            weight_list = np.concatenate([weight_list, weights[i,j,:]])
-            for k in range(bins):
-                bin_ids = []
-                materials = session.query(Material).filter(Material.run_id == run_id, Material.methane_loading_bin == i,
-                    Material.surface_area_bin == j, Material.void_fraction_bin == k,
-                    Material.dummy_test_result.in_(['none', 'pass'])).all()
-                for material in materials:
-                    bin_ids.append(material.id)
-                id_list = id_list + [bin_ids]
+    is selected, a parent is randomly-selected from those materials within that bin.
+    """
+    # Each bin is counted...
+    bins_and_counts = session.query(func.count(Material.id), Material.methane_loading_bin,
+        Material.surface_area_bin, Material.void_fraction_bin).filter(
+        Material.run_id == run_id).group_by(Material.methane_loading_bin,
+        Material.surface_area_bin, Material.void_fraction_bin).all()[1:]
```

Comment:
The wrapping is really confusing here. Maybe something like:

bins_and_counts = session.query(
      func.count(Material.id), 
      Material.methane_loading_bin, Material.surface_area_bin, Material.void_fraction_bin
    )
    .filter(Material.run_id == run_id)                                                            \
    .group_by(Material.methane_loading_bin,Material.surface_area_bin, Material.void_fraction_bin) \
    .all()[1:]